src/context/products/productContext.js

```diff
@@ -1,30 +1,50 @@
-import React, { useState, createContext, useContext, useEffect,useReducer } from "react";
-import { fetchProducts } from "../../services/product_services";
-import {reducer} from "./productReducer"
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useReducer,
+} from "react";
+import {fetchProducts,reducer,getSortedData,getFilterdData} from "./index"
 
 const ProductContext = createContext();
 
 const ProductContextProvider = ({ children }) => {
-  const intialState = {
+  let intialState = {
     products: [],
+    sortBy: null,
+    inStock: false,
+    fastDelivery: false,
+    brands: [],
+    selectedBrands:{},
+    ratings : null,
+    range: 5000,
+    categories: [],
+    selectedCategories : {}
+
   };
 
-  const [{ products }, dispatch] = useReducer(reducer, intialState);
+  const [state, dispatch] = useReducer(reducer, intialState);
+
+  let sortedData = getSortedData(state.products,state);
+  let filteredData = getFilterdData(sortedData,state);
 
   useEffect(() => {
     (async function () {
       const { data, success, message } = await fetchProducts();
 
       if (success) {
-        dispatch({ type: "FETCH_DATA", payload: { products: data } });
+        dispatch({
+          type: "FETCH_DATA",
+          payload: { products: data.products, categories: data.categories ,brands:data.brands},
+        });
       } else {
         //TODO : error handling
       }
     })();
   }, []);
 
```

Comment:
I am not sure if this useEffect should be used here or not, I think this would break the rules of hooks, but i am not sure, do let me know if i am right or wrong here

Reply:
It won't !!